File: src/atcoder/dp/ProblemE.java
package atcoder.dp;

import java.util.Scanner;

public class ProblemE {
    public static final int MAX_COST = 1000;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int w = scanner.nextInt();
        int[] dp = new int[n * MAX_COST + 1];
        int s = 0;
        for (int i = 0; i < n; i++) {
            int wi = scanner.nextInt();
            int vi = scanner.nextInt();
            s += vi;
            for (int j = s, k = s - vi; k > 0; j--, k--) {
                if (dp[k] != 0) {
                    int val = dp[k] + wi;
                    if (val <= w) {
                        if (dp[j] == 0) dp[j] = val;
                        else dp[j] = Math.min(dp[j], val);
                    }
                }
            }
            if (dp[vi] == 0) dp[vi] = wi;
            else dp[vi] = Math.min(dp[vi], wi);
        }
        for (int i = s; i >= 0; i--) {
            if (dp[i] != 0) {
                System.out.println(i);
                return;
            }
        }
        System.out.println(0);
    }
}

File: src/codeforces/contest1791/ProblemE.kt
@file:Suppress("DuplicatedCode")

package codeforces.contest1791

import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.util.*
import kotlin.math.absoluteValue

fun main(args: Array<String>) {
    val tests = nextInt()
    for (t in 0 until tests) {
        val n = nextInt()
        val arr = nextInts(n)
        println(solve(arr))
    }
}

fun solve(arr: IntArray): Long {
    val shouldBeOneNegative = arr.count { it < 0 } % 2 == 1
    var totalSum = arr.sumOf { it.toLong().absoluteValue }
    if (shouldBeOneNegative) totalSum -= 2 * arr.minOf { it.absoluteValue }
    return totalSum
}


private class MyReader(inputStream: InputStream) {
    private val reader = BufferedReader(InputStreamReader(inputStream))
    private var tokenizer: StringTokenizer? = null

    fun next(): String {
        while (true) {
            tokenizer.let {
                if (it == null || !it.hasMoreTokens()) {
                    tokenizer = StringTokenizer(reader.readLine())
                } else {
                    return it.nextToken()
                }
            }
        }
    }
}

private val reader = MyReader(System.`in`)
private fun next() = reader.next()
private fun nextInt() = next().toInt()
private fun nextLong() = next().toLong()
private fun nextDouble() = next().toDouble()
private fun nextStrings(n: Int) = Array(n) { next() }
private fun nextInts(n: Int) = IntArray(n) { nextInt() }
private fun nextLongs(n: Int) = LongArray(n) { nextLong() }
private fun nextDoubles(n: Int) = DoubleArray(n) { nextDouble() }